ceph/ceph_admin/apply_spec.py

```diff
@@ -0,0 +1,74 @@
+"""
+Represent the ceph orch CLI action 'apply' using spec file.
+
+Module to deploy ceph service(s) using orchestration command
+"ceph orch apply -i spec.yaml"
+
+"""
+import logging
+import tempfile
+from typing import Dict
+
+from .common import config_dict_to_string
+from .helper import PrepareSpec
+from .typing_ import ServiceProtocol
+
+LOG = logging.getLogger()
+
+
+class ApplySpec:
+    """Add apply command support to provision service using spec file."""
+
+    def apply_spec(self: ServiceProtocol, config: Dict) -> None:
+        """
+        Execute the apply_spec method using the object's service name and provided input.
+
+        Args:
+            config:     Key/value pairs passed from the test suite.
+                        base_cmd_args   - key/value pairs to set for base command
+                        specs           - service specifications.
+
+        Example:
+            config:
+                command: apply_spec
+                service: orch
+                base_cmd_args:          # arguments to ceph orch
+                    concise: true
+                    verbose: true
+                specs:
+                 - service_type: host
+                   attach_ip_address: true
+                   labels: apply-all-labels
+                   nodes:
+                     - node2
+                     - node3
+        """
+        base_cmd = ["ceph", "orch"]
+
+        if config.get("base_cmd_args"):
+            base_cmd_args_str = config_dict_to_string(config.get("base_cmd_args"))
+            base_cmd.append(base_cmd_args_str)
+
+        base_cmd.append("apply -i")
+        specs = config["specs"]
+
+        spec_cls = PrepareSpec(cluster=self.cluster, specs=specs)
+        spec_content = spec_cls.collect_spec()
+        LOG.info(f"Spec yaml file content:\n{spec_content}")
+
+        # Create spec yaml file
+        temp_file = tempfile.NamedTemporaryFile(suffix=".yaml")
+        base_cmd.append(temp_file.name)
+        spec_file = self.installer.node.remote_file(
+            sudo=True, file_name=temp_file.name, file_mode="w"
+        )
+        spec_file.write(spec_content)
+        spec_file.flush()
+
+        out, err = self.shell(
+            args=base_cmd,
+            base_cmd_args={"mount": "/tmp:/tmp"},
+        )
+
+        LOG.info(f"apply-spec command response :\n{out}")
+        # todo: add verification part
```

Comment:
Let's return `out, err` and allow the overriding methods process the verification. Doing it later allows more control.

Reply:
will include this when verification is added.